docs: improve cluster upgrade documentation TDE-1562 (#1047)
linz/topo-workflows: docs/infrastructure/cluster-upgrade.md

   ```bash
   aws eks wait nodegroup-active --cluster-name=Workflows --nodegroup-name="$node_group_name"
   ```

### k8s components

#### Upgrade `cdk8s-plus`

**It is a [good idea](https://cdk8s.io/docs/latest/plus/#i-operate-kubernetes-version-1xx-which-cdk8s-library-should-i-be-using) to check if a `cdk8s-plus` version matches the Kubernetes version you want to upgrade to.**

> If there is a version matching to the Kubernetes version to upgrade to, upgrade CDK8s-plus before proceeding to upgrade Kubernetes steps.

Reviewer: ```suggestion
If there is a version matching the Kubernetes version to upgrade to, upgrade CDK8s-plus before proceeding to upgrade Kubernetes steps.
```

1. Install the new version

   ```bash
   npm install --save-dev cdk8s-plus-27
   ```

2. Remove the previous version

   ```bash
   npm rm cdk8s-plus-26
   ```

> If there is no version matching, keep the version installed and proceed to upgrade Kubernetes steps.

Reviewer 2: "Keep the version installed" - may lead to keeping an outdated version for too long?
```suggestion
If there is no matching version, upgrade to the highest suitable version (skip this step if no higher version is available).
```

#### Upgrade components

For each of the component to upgrade:

- Use a compatible version based on Kubernetes version.
- Create a GitHub Pull Request so the upgrade can be done by the CI/CD pipeline that deploys these changes in production using `cdk8s` and `kubectl`.

##### Karpenter

- Update Helm values in `infra/charts/karpenter.ts`
- If API changed from last version, refers to [`karpenter` component documentation](components/karpenter.md) to see how to import new CRDs classes.

##### Fluent Bit

- Update Helm values in `infra/charts/fluentbit.ts`
- More information in [our fluentbit documentation](components/fluentbit.md)

##### Event Exporter

> - Update Helm values in `infra/charts/event.exporer.ts`

Author: exporTer might be better!

##### Non Production

The process is different in a Dev/NonProd cluster as the deployment will be manual:

1. Generate the kubernetes configuration yaml into `dist/`

   ```shell
   npx cdk8s synth
   ```

2. Apply the generated yaml files

   ```shell
   kubectl apply --filename=dist/
   ```

##### Clean-up

You can safely apply updated YAML manifests on top of the existing deployment, however if some CRDs have changed, for example in a Karpenter upgrade we've seen `Provisioner` -> `NodePool`, Helm does not delete the old CRDs. To avoid potential issues, like confusion from orphaned CRDs, it is a good thing to delete these old CRDs. After deployment:

```shell
kubectl get crds

> kubectl deleted crd <old-crd-name>

Reviewer 2: ```suggestion
kubectl delete crd <old-crd-name-1> <old-crd-name-n>
```

Reviewer 2: it _might_ be worthwhile pointing out that you _can_ delete multiple crds at once.
Either way, `deleted` has a type-o. :-)
